File: app/src/main/java/com/akrapovic/soundkit/community/service/ConnectReadyObserver.kt
package com.akrapovic.soundkit.community.service

import com.akrapovic.soundkit.community.domain.ConnectionState
import com.akrapovic.soundkit.community.domain.ValveState

/**
 * Detects the first moment a BLE link is ready for drive-mode automation within a session.
 * Valve state changes while connected must not count as a new connect-ready event.
 */
internal object ConnectReadyObserver {
    data class Transition(
        val isConnectReady: Boolean,
        val becameReady: Boolean,
        val disconnected: Boolean,
    )

    fun evaluate(
        connection: ConnectionState,
        valve: ValveState,
        notReady: String?,
        wasConnectReady: Boolean,
    ): Transition {
        val disconnected = connection is ConnectionState.Disconnected ||
            connection is ConnectionState.Error
        val isConnectReady = connection is ConnectionState.Connected &&
            valve != ValveState.Unknown &&
            notReady == null
        val becameReady = isConnectReady && !wasConnectReady
        return Transition(isConnectReady, becameReady, disconnected)
    }
}

File: ios/SoundKitCommunity/Domain/ConnectReadyObserver.swift
import Foundation

enum ConnectReadyObserver {
    struct Transition {
        let isConnectReady: Bool
        let becameReady: Bool
        let disconnected: Bool
    }

    static func evaluate(
        isConnected: Bool,
        valve: ValveState,
        receiverNotReady: Bool,
        wasConnectReady: Bool
    ) -> Transition {
        let disconnected = !isConnected
        let isConnectReady = isConnected && valve != .unknown && !receiverNotReady
        let becameReady = isConnectReady && !wasConnectReady
        return Transition(isConnectReady: isConnectReady, becameReady: becameReady, disconnected: disconnected)
    }
}
